File: cubelet_file.cc
#include <cassert>
#include <cstdio>
#include <cstring>
#include "cubelet_file.h"
#include "wavelet_compress.pb.h"

#ifdef _WIN32
#include <windows.h>
#include <io.h>
#include <fcntl.h>
#include <sys/stat.h>
#else
#include <unistd.h>
#include <sys/types.h>
#include <errno.h>
#endif

// 20 character header line
static const char *CUBELET_HEADER = "SCU cubelets 1.0\n";
#define HEADER_SIZE 64

/**
   Serialize a protocol buffer message to the given file.
   If writeLength is true, first write the length of the message as 
   a 4 byte int.
   Return the length of the message.
   On error, write message to stderr and return 0.
*/
static int writeProtobuf(FILE *outf, google::protobuf::Message *message,
                         bool writeLength = true);

static bool readProtobuf(FILE *inf, google::protobuf::Message *message,
                         uint32_t *length = NULL);

static bool fileExists(const char *filename);


// This is designed to be 64 bytes
struct CubeletHeader {
  char headerLine[24];

  char unused[40];

  CubeletHeader() {
    assert(sizeof(CubeletHeader) == HEADER_SIZE);
    strncpy(headerLine, CUBELET_HEADER, 24);
    headerLine[23] = 0;  // null-terminate it just in case

    memset(unused, 0, sizeof unused);
  }
};


/**
   Serialize a protocol buffer message to the given file.
   If writeLength is true, first write the length of the message as 
   a 4 byte int.
   Return the length of the message.
   On error, write message to stderr and return 0.
*/
static int writeProtobuf(FILE *outf, google::protobuf::Message *message,
                         bool writeLength) {
                          
  uint32_t codedLen = message->ByteSize();
  if (writeLength)
    fwrite(&codedLen, sizeof codedLen, 1, outf);

  // encode the metadata
  char *codedBuf = new char[codedLen];
  assert(codedBuf);
  if (!message->SerializeToArray(codedBuf, codedLen)) {
    fprintf(stderr, "Failed to encode metadata\n");
    return 0;
  }

  // write the metadata to the file
  if (fwrite(codedBuf, 1, codedLen, outf) != codedLen) {
    fprintf(stderr, "Failed to write encoded metadata to file\n");
    return 0;
  }
  delete[] codedBuf;

  return codedLen;
}  


static bool readProtobuf(FILE *inf, google::protobuf::Message *message,
                         uint32_t *length) {

  // read the length
  uint32_t codedLen;
  if (fread(&codedLen, sizeof codedLen, 1, inf) != 1) {
    // this will happen naturally at EOF
    // fprintf(stderr, "Failed to read message buffer size\n");
    return false;
  }

  // return the length, if the caller is interested
  if (length) *length = codedLen;

  // this is a bit of a hack, but if the length is 0xFFFFFFFF,
  // consider it invalid and return false. This marks EOF.
  if (codedLen == 0xFFFFFFFF) return false;

  // allocate a buffer
  // XXX might want to have a small (32 bytes or so) buffer array on the
  // stack as a possible shortcut
  char *codedBuf = new char[codedLen];
  assert(codedBuf);

  // read the encoded data from the file
  if (fread(codedBuf, 1, codedLen, inf) != codedLen) {
    fprintf(stderr, "Failed to read metadata\n");
    return false;
  }

  // decode the metadata
  if (!message->ParseFromArray(codedBuf, codedLen)) {
    fprintf(stderr, "Failed to decode metadata\n");
    return false;
  }

  delete[] codedBuf;
  
  return true;
}


static bool fileExists(const char *filename) {
  FILE *inf = fopen(filename, "r");
  if (inf) {
    fclose(inf);
    return true;
  } else {
    return false;
  }
}


bool isCubeletFile(const char *filename) {
  if (!filename || !strcmp("-", filename)) return false;

  CubeletStreamReader reader;
  if (reader.open(filename, true)) {
    reader.close();
    return true;
  } else {
    return false;
  }
}


bool CubeletStreamWriter::open(const char *filename, bool append) {
  // basic assumption in a 32-bit world
  assert(sizeof(unsigned) == 4);

  if (outf) close();
  index.clear_cubelets();

  if (filename && strcmp(filename, "-")) {

    if (append) {

      // if the file doesn't exist, just create it like normal
      if (fileExists(filename)) {

	// if it exists but is not a cubelet file, complain and fail
	if (!isCubeletFile(filename)) {
	  fprintf(stderr, "Error: \"%s\" is not a cubelet file.\n", filename);
	  return false;
	}

	// open the file for writing, positioned after the last cubelet,
	// with the footer data truncated.
	return openAfterLastCubelet(filename);
      }

    }

    outf = fopen(filename, "wb");
    if (!outf) {
      fprintf(stderr, "Error: cannot open \"%s\" for writing.\n", filename);
      return false;
    }
  } else {
    outf = stdout;
  }

  CubeletHeader header;

  size_t bytesWritten = fwrite(&header, 1, sizeof(header), outf);
  if (bytesWritten != sizeof(header)) {
    fprintf(stderr, "Failed to write cubelet file header\n");
    if (outf != stdout) fclose(outf);
    outf = NULL;
    return false;
  }
  
  return true;
}


bool CubeletStreamWriter::openAfterLastCubelet(const char *filename) {
  
  // find the last cubelet in the file
  CubeletStreamReader reader;
  if (!reader.open(filename)) return false;

  std::vector<Cube> cubelets;
  if (!reader.listCubelets(cubelets)) return false;

  off_t offsetAfterLastCubelet = HEADER_SIZE;
  for (size_t i=0; i < cubelets.size(); i++) {
    off_t offsetAfter = cubelets[i].dataFileOffset + 
      cubelets[i].getSizeInBytes();
    if (offsetAfter > offsetAfterLastCubelet)
      offsetAfterLastCubelet = offsetAfter;

    // add the the list of cubelets in this file
    CubeletBuffer *buffer = index.add_cubelets();
    cubelets[i].copyToCubeletBuffer(buffer);
  }
  reader.close();

  // truncate the file after the last cubelet
#ifndef _WIN32
  if (truncate(filename, offsetAfterLastCubelet)) {
    fprintf(stderr, "Error truncating \"%s\": %s\n", filename, strerror(errno));
    return false;
  }
#else
  int fd;
  if (_sopen_s(&fd, filename, _O_BINARY | _O_WRONLY, _SH_DENYRW,
               _S_IREAD | _S_IWRITE)) {
    fprintf(stderr, "Error opening \"%s\" for truncation: %s\n",
            filename, strerror(errno));
    return false;
  }
  if (_chsize_s(fd, offsetAfterLastCubelet)) {
    fprintf(stderr, "Error truncating \"%s\": %s\n",
            filename, strerror(errno));
    return false;
  }
#endif

  outf = fopen(filename, "a");
  if (!outf) {
    fprintf(stderr, "Failed to open \"%s\" for writing.\n", filename);
    return false;
  }

  // printf("open to offset %d\n", (int)ftell(outf));

  return true;
}



bool CubeletStreamWriter::addCubelet(const Cube *cubelet) {
  if (!outf) {
    fprintf(stderr, "Cannot add cubelets. Stream is not open yet.\n");
    return false;
  }
  
  CubeletBuffer *buffer = index.add_cubelets();

  assert(cubelet->size > scu_wavelet::int3(0,0,0));

  cubelet->copyToCubeletBuffer(buffer);

  unsigned dataByteCount = 0;

  if (cubelet->data_) {
    dataByteCount = cubelet->getSizeInBytes();
    buffer->set_byte_count(dataByteCount);
  }

  writeProtobuf(outf, buffer);

  // save the position of the data so it can be included in the index
  buffer->set_data_file_offset(ftell(outf));

  // XXX should cubelet->dataFileOffset be set? const will have to be
  // casted off.

  /*
  printf("Write %d cubelet bytes at offset %llu\n", dataByteCount,
         (long long unsigned) buffer->data_file_offset());
  */
  
  // if the data hasn't been set, don't output any data
  if (cubelet->data_) {
    if (!cubelet->writeToFile(outf)) {
      fprintf(stderr, "Failed to write cubelet data to file\n");
      return false;
    }
  }
  
  return true;
}


bool CubeletStreamWriter::close(bool noFooter) {
  // If the file isn't open do nothing.
  if (!outf) return true;

  if (!noFooter) {
    int32_t eofMarker = -1; // set all bits
    fwrite(&eofMarker, sizeof eofMarker, 1, outf);
  
    // write the cubelet index
    int32_t indexLen = writeProtobuf(outf, &index, false);

    if (fwrite(&indexLen, sizeof indexLen, 1, outf) != 1) {
      fprintf(stderr, "Failed to write index size in the file footer.\n");
    }

    // write special tag to show the footer is valid
    fwrite("cube", 1, 4, outf);
  }

  if (outf != stdout) fclose(outf);

  outf = NULL;

  return true;
}


bool CubeletStreamReader::open(const char *filename, bool quiet) {
  if (inf) close();

  if (filename && strcmp(filename, "-")) {
    inf = fopen(filename, "rb");
    if (!inf) {
      if (!quiet)
	fprintf(stderr, "Cannot open \"%s\" for reading.\n", filename);
      return false;
    }
  } else {
    inf = stdin;
  }

  CubeletHeader header;

  // read the header
  if (fread(&header, sizeof header, 1, inf) != 1) {
    if (!quiet)
      fprintf(stderr, "Failed to read header in \"%s\".\n", filename);
    if (inf != stdout) fclose(inf);
    inf = NULL;
    return false;
  }

  if (strcmp(header.headerLine, CUBELET_HEADER)) {
    if (!quiet)
      fprintf(stderr, "Input file \"%s\" is not a cubelet file.\n", filename);
    if (inf != stdout) fclose(inf);
    inf = NULL;
    return false;
  }

  dataSizeBytes = 0;
  dataHasBeenRead = true;
  eofReached = false;

  return true;
}


bool CubeletStreamReader::reset() {
  // reset on a closed file might as well be a no-op
  if (!inf) return true;

  if (fseek(inf, sizeof(CubeletHeader), SEEK_SET)) {
    fprintf(stderr, "Failed to jump back to the top of the cubelet file.\n");
    return false;
  }
  
  dataSizeBytes = 0;
  dataHasBeenRead = true;
  eofReached = false;

  return true;
}


bool CubeletStreamReader::next(Cube *cube) {
  // if already EOF, don't read more
  if (eofReached) return false;

  // if the data for this cubelet has not been read, seek past it
  if (!dataHasBeenRead) {
    if (fseek(inf, dataSizeBytes, SEEK_CUR)) {
      eofReached = true;
      return false;
    }
  }

  CubeletBuffer buf;

  // if whatever comes next is not a parseable cubelet, we're done
  if (!readProtobuf(inf, &buf)) {
    eofReached = true;
    return false;
  }

  cube->copyFromCubeletBuffer(&buf);
  cube->dataFileOffset = ftell(inf);
  cube->data_ = NULL;

  dataSizeBytes = buf.byte_count();
  dataHasBeenRead = false;

  currentCubeSize = cube->size;

  return true;
}


bool CubeletStreamReader::find(Cube *cube, int3 id) {
  if (!reset()) return false;
  
  Cube tmpCube;
  while (true) {
    if (!next(&tmpCube)) return false;
    
    if (id < int3(0,0,0) || id == tmpCube.parentOffset) {
      *cube = tmpCube;
      return true;
    }
  }
}
  

void *CubeletStreamReader::getRawData(void *data) {

  if (dataSizeBytes == 0) return NULL;
  
  if (!data)
    data = malloc(dataSizeBytes);

  // the data can only be read once for each cubelet to avoid buffering
  // or seeking backwards
  if (dataHasBeenRead) {
    fprintf(stderr, "Error: cannot read the data for one cubelet twice.\n");
    return NULL;
  }

  if (fread(data, 1, dataSizeBytes, inf) != dataSizeBytes) {
    fprintf(stderr, "Failed to read cubelet content data.\n");
    return NULL;
  }

  dataHasBeenRead = true;

  return data;
}

/**
   Read the data for the curent cube.
   If cube->dataFileOffset is not set, this fails and returns false.

   data is compressed:
     storage is not allocated: allocate and store
     storage is allocated: assume it's the size of the compressed data?
   data is not compressed
     storage is not allocated: allocate and store
     storage is allocated: 

*/
bool CubeletStreamReader::getCubeData(Cube *cube) {

  if (cube->dataFileOffset == 0) {
    fprintf(stderr, "Cannot read cubelet %s; dataFileOffset not set.\n",
	    cube->getId());
    return false;
  }

  fseek(inf, cube->dataFileOffset, SEEK_SET);
  dataHasBeenRead = false;
  dataSizeBytes = cube->getSizeInBytes();
  if (!cube->data_) cube->allocate();

  // if the data is compressed, just store it directly
  if (cube->isWaveletCompressed) {
    cube->data_ = getRawData(cube->data_);
    if (!cube->data_) return false;
  }

  // if it isn't compressed, read it into the cube, resizing if necessary
  else {

    // if there is no funny business with the size, no change needed
    if (currentCubeSize == cube->size &&
        currentCubeSize == cube->totalSize) {
      cube->data_ = getRawData(cube->data_);
      if (!cube->data_) return false;
    }

    // translation might be needed
    else {

      assert(currentCubeSize <= cube->size);

      void *rawData = getRawData(NULL);
      if (!rawData) return false;
    
      switch (cube->datatype) {
      case WAVELET_DATA_UINT8:
        copyIntoTypedCube((CubeByte*)cube, rawData, currentCubeSize);
        break;
      case WAVELET_DATA_INT32:
        copyIntoTypedCube((CubeInt*)cube, rawData, currentCubeSize);
        break;
      case WAVELET_DATA_FLOAT32:
        copyIntoTypedCube((CubeFloat*)cube, rawData, currentCubeSize);
        break;
      default:
        fprintf(stderr, "Error reading cubelet data: unrecognized type id %d\n",
                cube->datatype);
        return false;
      }

      free(rawData);
    }
  }

  return true;
}


bool CubeletStreamReader::listCubelets(std::vector<Cube> &cubelets) {

  reset();
  cubelets.clear();
  Cube cube;
  
  while (next(&cube)) {
    cubelets.push_back(cube);
  }

  return true;
}

    
void CubeletStreamReader::close() {
  if (inf && inf != stdout) fclose(inf);
  inf = NULL;
}


/*
  Might want to use this:


  Borrowed from https://cxwangyi.wordpress.com/2010/07/20/encoding-and-decoding-of-the-varint32-type-defined-by-google-protocol-buffers/
  which borrowed from http://protobuf.googlecode.com/svn/trunk/src/google/protobuf/io/coded_stream.cc

#include <google/protobuf/io/coded_stream.h>
#include "../base/common.hh"
 
using namespace std;
using namespace google::protobuf::io;
 
inline const uint8* ReadVarint32FromArray(const uint8* buffer, uint32* value) {
  static const int kMaxVarintBytes = 10;
  static const int kMaxVarint32Bytes = 5;
 
  // Fast path:  We have enough bytes left in the buffer to guarantee that
  // this read won't cross the end, so we can skip the checks.
  const uint8* ptr = buffer;
  uint32 b;
  uint32 result;
 
  b = *(ptr++); result  = (b & 0x7F)      ; if (!(b & 0x80)) goto done;
  b = *(ptr++); result |= (b & 0x7F) <<  7; if (!(b & 0x80)) goto done;
  b = *(ptr++); result |= (b & 0x7F) << 14; if (!(b & 0x80)) goto done;
  b = *(ptr++); result |= (b & 0x7F) << 21; if (!(b & 0x80)) goto done;
  b = *(ptr++); result |=  b         << 28; if (!(b & 0x80)) goto done;
 
  // If the input is larger than 32 bits, we still need to read it all
  // and discard the high-order bits.
  for (int i = 0; i < kMaxVarintBytes - kMaxVarint32Bytes; i++) {
    b = *(ptr++); if (!(b & 0x80)) goto done;
  }
 
  // We have overrun the maximum size of a varint (10 bytes).  Assume
  // the data is corrupt.
  return NULL;
 
 done:
  *value = result;
  return ptr;
}
*/

File: CubeletFile.java
import java.nio.charset.Charset;
import java.io.*;
import java.util.ArrayList;
import com.google.protobuf.InvalidProtocolBufferException;

/**
   Java code for handling cubelet files.

   This uses the Google Protocol Buffer API, so when compiling
   or running it, you'll need "protobuf.jar" in the classpath.
   
   On Linux, if the libprotobuf-java package is installed,
   it will probably be in /usr/share/java/protobuf.jar.
   On Windows, a good place to store it would be
   at ./protobuf-2.6.0/protobuf.jar
*/
public class CubeletFile {

  private static final String HEADER_LINE = "SCU cubelets 1.0\n";
  private static final int HEADER_SIZE = 64;

  private static Charset ASCII;

  static {
    ASCII = Charset.forName("US-ASCII");
  }


  // test function - list the cubelets in a file
  public static void main(String args[]) throws Exception {
    if (args.length != 1) printHelp();

    String filename = args[0];
    if (filename.equals("-")) filename = null;
    
    listCubelets(filename);
  }


  // test function - write simple cubelet file
  public static void main2(String[] args) throws Exception {
    Cubelet cube = new Cubelet();
    float[] data = new float[100*100*100];

    String filename = args[0];
    if (filename.equals("-")) filename = null;

    Writer out = new Writer();
    out.open(filename);

    cube.datatype = Cubelet.DataType.FLOAT32;
    cube.floatData = data;

    cube.setSize(4, 5, 6);
    cube.setOffset(100, 101, 102);
    for (int i=0; i < 4*5*6; i++) data[i] = i*.25f;
    out.addCubelet(cube);
    
    cube.setSize(7, 8, 9);
    cube.setOffset(200, 201, 202);
    for (int i=0; i < 7*8*9; i++) data[i] = i*.5f;
    out.addCubelet(cube);

    out.close();
  }


  public static void listCubelets(String filename) throws Exception {
    Reader reader = new Reader();

    try {
      reader.open(filename);
    } catch (CubeletFormatException e) {
      System.out.println(e.getMessage());
      return;
    } catch (FileNotFoundException e) {
      System.out.println("\"" + filename + "\" not found.");
      return;
    } catch (IOException e3) {
      System.out.println("Could not open cubelet file \"" + filename + 
                         "\":" + e3);
      return;
    }

    Cubelet cube = new Cubelet();

    while (reader.next(cube)) {
      System.out.println(cube.toString());

      reader.getData(cube);

      if (cube.datatype == Cubelet.DataType.FLOAT32) {
        for (int y=0; y < cube.height; y++) {
          for (int x=0; x < cube.width; x++) {
            System.out.printf("%6.2f", cube.floatData[y*cube.width + x]);
          }
          System.out.println();
        }
      } else if (cube.datatype == Cubelet.DataType.UINT8) {
        for (int y=0; y < cube.height; y++) {
          for (int x=0; x < cube.width; x++) {
            System.out.printf("%4d", 0xff & cube.byteData[y*cube.width + x]);
          }
          System.out.println();
        }
      } else if (cube.datatype == Cubelet.DataType.INT32) {
        for (int y=0; y < cube.height; y++) {
          for (int x=0; x < cube.width; x++) {
            System.out.printf("%7d", 0xff & cube.intData[y*cube.width + x]);
          }
          System.out.println();
        }
      }
        
    }

    reader.close();
  }


  private static void printHelp() {
    System.out.println("\n  CubeletFile <file>\n" +
                       "  List the contents of the file.\n");
    System.exit(1);
  }
  
    
  /** Read a little endian long from a byte array. */
  public static long longFromByteArray(byte[] array, int offset) {
    long value = 0;
    for (int i=7; i >= 0; i--) {
      value = (value << 8) | (array[offset+i] & 0xff);
    }
    return value;
  }
    
  /** Read a little endian int from a byte array. */
  public static int intFromByteArray(final byte[] array, int offset) {
    int value = 0;
    for (int i=3; i >= 0; i--) {
      value = (value << 8) | (array[offset+i] & 0xff);
    }
    return value;
  }
    
  /** Write a little endian int to a byte array. */
  public static void intToByteArray(int value, byte[] array, int offset) {
    for (int i=0; i < 4; i++) {
      array[offset+i] = (byte) value;
      value >>>= 8;
    }
  }

  static class CubeletFormatException extends Exception {
    public CubeletFormatException(String filename) {
      super("\"" + filename + "\" is not a cubelet file.");
    }
  }


  static class CubeletStreamingException extends Exception {
    public CubeletStreamingException() {}
  }


  public static class Reader {
    private InputStream in;
    private long cubeletIndexOffset;
    private int dataSizeBytes = 0;
    private Cubelet.DataType currentCubeletDatatype;
    private boolean dataHasBeenRead = true;
    private boolean eofReached = false;

    // re-use a buffer for reading encoded messages
    // must be at least 4 bytes long, but will be grown as necessary
    private byte[] readBuffer = new byte[32];

    com.google.protobuf.Parser<WaveletCompress.CubeletBuffer> cubeletParser;


    public Reader() {
      cubeletParser = WaveletCompress.CubeletBuffer.getDefaultInstance()
        .getParserForType();
    }


    /** Opens the given file for reading.
        Throws CubeletFormatException if the file is not a cubelet file.
        Throws FileNotFoundException if the file is not found.
     */
    public void open(String filename)
      throws IOException, FileNotFoundException, CubeletFormatException {

      if (in != null) in.close();
      
      if (filename == null) {
        in = System.in;
      } else {
        in = new BufferedInputStream
          (new FileInputStream(filename));
      }

      byte[] headerBuf = new byte[HEADER_SIZE];
      int bytesRead = in.read(headerBuf);
      if (bytesRead != HEADER_SIZE) {
        close();
        throw new IOException("Failed to read header");
      }

      String headerLine = new String(headerBuf, 0, HEADER_LINE.length(), ASCII);

      if (!headerLine.equals(HEADER_LINE)) {
        close();
        throw new CubeletFormatException(filename);
      }

      cubeletIndexOffset = longFromByteArray(headerBuf, 24);

      dataSizeBytes = 0;
      dataHasBeenRead = true;
      eofReached = false;
    }
    
    
    /** Reads a cubelet from the file, stores its metadata in 'cube',
        and returns true.
        Returns false when EOF reached or if the file has not been opened.
        Throws IOException on data format error or other IO error. */
    public boolean next(Cubelet cube) throws IOException {
      if (in == null || eofReached) return false;
      
      // if the data for this chunk hasn't been read, skip over it before
      // trying to read the next chunk.
      if (!dataHasBeenRead) {
        // skip() doesn't always skip the requested number of bytes,
        // so it must be called in a loop
        long bytesToSkip = dataSizeBytes;
        while (bytesToSkip > 0) {
          long bytesSkipped = in.skip(bytesToSkip);
          bytesToSkip -= bytesSkipped;
        }
      }

      WaveletCompress.CubeletBuffer cubeBuf = readCubeletBuffer();
      if (cubeBuf == null) {
        eofReached = true;
        return false;
      }

      cube.setSize(cubeBuf.getWidth(), cubeBuf.getHeight(), cubeBuf.getDepth());
      cube.setOffset(cubeBuf.getXOffset(), cubeBuf.getYOffset(),
                     cubeBuf.getZOffset());

      switch (cubeBuf.getDataType()) {
      case UINT8: cube.datatype = Cubelet.DataType.UINT8; break;
      case FLOAT32: cube.datatype = Cubelet.DataType.FLOAT32; break;
      case INT32: cube.datatype = Cubelet.DataType.INT32; break;
      }

      switch (cubeBuf.getCompressionAlgorithm()) {
      case NONE: cube.compressionAlg = Cubelet.CompressionAlg.NONE; break;
      case ZLIB: cube.compressionAlg = Cubelet.CompressionAlg.ZLIB; break;
      case WAVELET: cube.compressionAlg = Cubelet.CompressionAlg.WAVELET; break;
      }

      cube.maxPossibleValue = cubeBuf.getMaximumValue();

      cube.dataFileOffset = 0;
      cube.byteData = null;
      cube.floatData = null;
      cube.intData = null;

      // save the size of the data for this buffer
      dataSizeBytes = cubeBuf.getByteCount();
      currentCubeletDatatype = cube.datatype;
      dataHasBeenRead = false;

      return true;
    }


    /** Get the data for the current cubelet.
        If the data is UINT8, store it in cube.byteData.
        If the data is FLOAT32, store it in cube.floatData.
        If the data is INT32, store it in cube.intData.
        Throws CubeletStreamingException if the data for this cubelet has
        already been read (the input data may be streaming, so we can't
        back up to read the data again).
    */
    public void getData(Cubelet cube)
      throws IOException, CubeletStreamingException {

      if (dataSizeBytes == 0) return;
      
      if (dataHasBeenRead)
        throw new CubeletStreamingException();
      
      byte[] bytes = new byte[dataSizeBytes];
      int bytesRead = in.read(bytes);
      if (bytesRead != dataSizeBytes)
        throw new IOException("Failed to read cubelet data");

      dataHasBeenRead = true;
      
      if (currentCubeletDatatype == Cubelet.DataType.UINT8) {
        cube.byteData = bytes;
      }

      else if (currentCubeletDatatype == Cubelet.DataType.FLOAT32) {
        int count = dataSizeBytes / 4;
        float[] floats = new float[count];

        for (int i=0; i < count; i++) {
          // convert the bytes to ints, then to floats
          int tmp = intFromByteArray(bytes, i*4);
          floats[i] = Float.intBitsToFloat(tmp);
        }
        
        cube.floatData = floats;
      }

      else if (currentCubeletDatatype == Cubelet.DataType.INT32) {
        int count = dataSizeBytes / 4;
        int[] ints = new int[count];

        for (int i=0; i < count; i++) {
          // convert the bytes to ints, then to floats
          ints[i] = intFromByteArray(bytes, i*4);
        }
        
        cube.intData = ints;
      }

    }
    

    public void close() {
      if (in != null && in != System.in) {
        try {
          in.close();
        } catch (IOException e) {}
      }
      in = null;
    }


    // make sure the read buffer is at least this large
    private void readBufferReserve(int size) {
      if (readBuffer.length < size) {
        int newSize = Math.max(size, readBuffer.length*2);
        readBuffer = new byte[newSize];
      }
    }

    /** Read a CubeletBuffer from the input stream. */
    private WaveletCompress.CubeletBuffer readCubeletBuffer()
      throws IOException {

      // start by reading 4-byte length field
      int bytesRead = in.read(readBuffer, 0, 4);

      // end of file, just return null
      if (bytesRead != 4) return null;

      // decode the length
      int encodedLength = intFromByteArray(readBuffer, 0);
      /*
      System.out.printf("encodedLength %02x %02x %02x %02x = %d\n",
                        readBuffer[0]&0xff, readBuffer[1]&0xff,
                        readBuffer[2]&0xff, readBuffer[3]&0xff,
                        encodedLength);
      */

      // length of 0xFFFFFFFF marks EOF 
      if (encodedLength == -1) return null;
      
      readBufferReserve(encodedLength);

      // read the encoded data
      bytesRead = in.read(readBuffer, 0, encodedLength);
      if (bytesRead != encodedLength) return null;

      WaveletCompress.CubeletBuffer cubeletBuffer = null;

      // decode the metadata
      try {
        cubeletBuffer = cubeletParser.parseFrom(readBuffer, 0, encodedLength);
      } catch (InvalidProtocolBufferException e) {
        throw new IOException("Failed to decode metadata");
      }

      return cubeletBuffer;
    }
  }


  /**
     Pass-through OutputStream that counts the number of bytes
     written to it.
  */
  public static class CountingOutputStream extends OutputStream {
    private OutputStream out;
    private long byteCount;

    public long getByteCount() {
      return byteCount;
    }

    public CountingOutputStream(OutputStream out) {
      super();
      this.out = out;
    }

    public void write(byte[] b) throws IOException {
      out.write(b);
      byteCount += b.length;
    }

    public void write(byte[] b, int off, int len) throws IOException {
      out.write(b, off,len);
      byteCount += len;
    }
     
    public void write(int b) throws IOException {
      out.write(b);
      byteCount++;
    }

    public void flush() throws IOException {
      out.flush();
    }

    public void close() throws IOException {
      out.flush();
      out.close();
    }
  }


  /**
     Write a cubelet file. 
  */
  public static class Writer {

    private boolean isStdout;
    private CountingOutputStream out;
    ArrayList<WaveletCompress.CubeletBuffer> index
      = new ArrayList<WaveletCompress.CubeletBuffer>();

    public int width=0, height=0, depth=0;

    /** Open the given file for writing.
        If filename is null, use System.out.
        Write the header.
    */
    public void open(String filename) throws IOException {
      if (out != null) close();
      
      if (filename != null) {
        isStdout = false;
        out = new CountingOutputStream
          (new BufferedOutputStream
           (new FileOutputStream(filename)));
      } else {
        out = new CountingOutputStream(System.out);
        isStdout = true;
      }
      
      byte[] headerBuf = new byte[HEADER_SIZE];

      // set the header line
      byte[] headerLine = HEADER_LINE.getBytes(ASCII);
      System.arraycopy(headerLine, 0, headerBuf, 0, headerLine.length);
      
      out.write(headerBuf);
    }


    /**
       Write a cubelet (and its data) to the output stream.
    */
    public void addCubelet(Cubelet cube) throws IOException {
      if (out == null) throw new NullPointerException("File not open");

      if (cube.datatype == null)
        throw new NullPointerException("cubelet data type not set");

      int pixelSize = 0;
      int count = cube.width * cube.height * cube.depth;
      byte[] bytes = null;

      WaveletCompress.CubeletBuffer.DataType datatype = null;

      switch (cube.datatype) {
      case UINT8:
        datatype = WaveletCompress.CubeletBuffer.DataType.UINT8;
        if (cube.byteData == null)
          throw new NullPointerException("cubelet data not set");
        bytes = cube.byteData;
        pixelSize = 1;
        break;

      case FLOAT32:
        datatype = WaveletCompress.CubeletBuffer.DataType.FLOAT32;
        if (cube.floatData == null)
          throw new NullPointerException("cubelet data not set");

        pixelSize = 4;
        bytes = new byte[count * pixelSize];
        for (int i=0; i < count; i++) {
          int tmp = Float.floatToIntBits(cube.floatData[i]);
          intToByteArray(tmp, bytes, i*4);
        }
        break;

      case INT32:
        datatype = WaveletCompress.CubeletBuffer.DataType.INT32;
        if (cube.intData == null)
          throw new NullPointerException("cubelet data not set");

        pixelSize = 4;
        bytes = new byte[count * pixelSize];
        for (int i=0; i < count; i++) {
          intToByteArray(cube.intData[i], bytes, i*4);
        }
        break;

      }

      WaveletCompress.CubeletBuffer cubeBuf
        = WaveletCompress.CubeletBuffer.newBuilder()
        .setWidth(cube.width)
        .setHeight(cube.height)
        .setDepth(cube.depth)
        .setXOffset(cube.xOffset)
        .setYOffset(cube.yOffset)
        .setZOffset(cube.zOffset)
        .setByteCount(pixelSize * count)
        .setDataType(datatype)
        .setMaximumValue(cube.maxPossibleValue)
        .build();

      // write the metadata
      writeProtobuf(cubeBuf);

      // save our current position in the output stream
      long cubeletFileOffset = out.getByteCount();

      cubeBuf = cubeBuf.toBuilder()
        .setDataFileOffset(cubeletFileOffset)
        .build();
      index.add(cubeBuf);
      // System.out.println("Write cubelet at offset " + cubeletFileOffset);

      // write the cubelet data
      out.write(bytes, 0, cubeBuf.getByteCount());
    }


    /**
       Writes a message to the output file, prefixed with its 4-byte length.
       Returns the number of bytes in the message.
    */
    private int writeProtobuf(com.google.protobuf.Message m)
      throws IOException {
      return writeProtobuf(m, true);
    }

                              
    /**
       Writes a message to the output file. If 'writeLength' is true,
       prefix the message with its 4-byte length.
       Returns the number of bytes in the message.
    */
    private int writeProtobuf(com.google.protobuf.Message m,
                              boolean writeLength)
      throws IOException {
      /*
        consider Parser<MessageType>.parseDelimitedFrom(InputStream),
        MessageLite.writeDelimitedTo(),
        CodedOutpuStream::writeVarint32ToArray
        (https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.coded_stream#CodedOutputStream.WriteVarint32)
        and ReadVarint32FromArray (saved in cubelet_file.cc)
      */

      // get the encoded length of the message
      int messageLen = m.getSerializedSize();
      
      // write the length as a 4-byte int
      if (writeLength) {
        byte[] lengthBuf = new byte[4];
        intToByteArray(messageLen, lengthBuf, 0);
        out.write(lengthBuf);
      }

      // write the encoded message
      m.writeTo(out);

      return messageLen;
    }


    /**
       Write the footer for the cubelet file and close the file.
    */
    public void close() throws IOException {
      if (out == null) return;

      // mark the end of cubelets with a length of 0xFFFFFFFF
      byte[] footerBuf = new byte[8];
      intToByteArray(-1, footerBuf, 0);
      out.write(footerBuf, 0, 4);

      // build the footer
      WaveletCompress.CubeletIndexBuffer.Builder indexBuilder
        = WaveletCompress.CubeletIndexBuffer.newBuilder();
      for (int i=0; i < index.size(); i++)
        indexBuilder.addCubelets(index.get(i));

      if (width > 0)  indexBuilder.setWidth(width);
      if (height > 0) indexBuilder.setHeight(height);
      if (depth > 0)  indexBuilder.setDepth(depth);
        
      WaveletCompress.CubeletIndexBuffer indexBuf = indexBuilder.build();
        
      // write the footer
      int indexLen = writeProtobuf(indexBuf, false);

      // at the end of the file, write an offset to the location of the
      // footer data, as well as a special byte sequence so we can
      // distinguish between a completed file and one that got truncated.
      intToByteArray(indexLen, footerBuf, 0);
      footerBuf[4] = 'c';
      footerBuf[5] = 'u';
      footerBuf[6] = 'b';
      footerBuf[7] = 'e';
      out.write(footerBuf);

      out.flush();

      if (!isStdout) out.close();

    }
  }
    
}

